service/controller/resource/workermigration/create.go

```diff
@@ -0,0 +1,496 @@
+
+	var dcList corev1alpha1.DrainerConfigList
+	err := r.ctrlClient.List(ctx, &dcList, o)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for _, dc := range dcList.Items {
+		err = r.ctrlClient.Delete(ctx, &dc)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	return nil
+}
+
+func (r *Resource) ensureAzureMachinePoolExists(ctx context.Context, cr providerv1alpha1.AzureConfig, vmSize string) (expcapzv1alpha3.AzureMachinePool, error) {
+	var azureMachinePool expcapzv1alpha3.AzureMachinePool
+
+	nsName := types.NamespacedName{
+		Name:      cr.GetName(),
+		Namespace: cr.GetNamespace(),
+	}
+	err := r.ctrlClient.Get(ctx, nsName, &azureMachinePool)
+	if err == nil {
+		// NOTE: Success return when AzureMachinePool CR already exists.
+		return azureMachinePool, nil
+	} else if errors.IsNotFound(err) {
+		// This is ok. CR gets created in a bit.
+	} else if err != nil {
+		return expcapzv1alpha3.AzureMachinePool{}, microerror.Mask(err)
+	}
+
+	var dockerDiskSizeGB int
+	var kubeletDiskSizeGB int
+	{
+		if key.WorkerCount(cr) > 0 {
+			dockerDiskSizeGB = cr.Spec.Azure.Workers[0].DockerVolumeSizeGB
+			kubeletDiskSizeGB = cr.Spec.Azure.Workers[0].KubeletVolumeSizeGB
+		}
+
+		if dockerDiskSizeGB <= 0 {
+			dockerDiskSizeGB = 100
+		}
+
+		if kubeletDiskSizeGB <= 0 {
+			kubeletDiskSizeGB = 100
+		}
+	}
+
+	// CR didn't exist so it's created here.
+	azureMachinePool = expcapzv1alpha3.AzureMachinePool{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "AzureMachinePool",
+			APIVersion: "exp.infrastructure.cluster.x-k8s.io/v1alpha3",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cr.Name,
+			Namespace: cr.Namespace, // TODO: Adjust for CR namespacing.
+			Labels: map[string]string{
+				label.AzureOperatorVersion:    key.OperatorVersion(&cr),
+				label.Cluster:                 key.ClusterID(&cr),
+				capiv1alpha3.ClusterLabelName: key.ClusterName(&cr),
+				label.MachinePool:             key.ClusterID(&cr),
+				label.Organization:            key.OrganizationID(&cr),
+				label.ReleaseVersion:          key.ReleaseVersion(&cr),
+			},
+		},
+		Spec: expcapzv1alpha3.AzureMachinePoolSpec{
+			Location: r.location,
+			Template: expcapzv1alpha3.AzureMachineTemplate{
+				DataDisks: []capzv1alpha3.DataDisk{
+					{
+						NameSuffix: "docker",
+						DiskSizeGB: int32(dockerDiskSizeGB),
+						Lun:        to.Int32Ptr(21),
+					},
+					{
+						NameSuffix: "kubelet",
+						DiskSizeGB: int32(kubeletDiskSizeGB),
+						Lun:        to.Int32Ptr(22),
+					},
+				},
+				SSHPublicKey: base64.StdEncoding.EncodeToString([]byte(key.AdminSSHKeyData(cr))),
+				VMSize:       vmSize,
+			},
+		},
+	}
+
+	err = r.ctrlClient.Create(ctx, &azureMachinePool)
+	if err != nil {
+		return expcapzv1alpha3.AzureMachinePool{}, microerror.Mask(err)
+	}
+
+	return azureMachinePool, nil
+}
+
+func (r *Resource) ensureDrainerConfigsExists(ctx context.Context, cr providerv1alpha1.AzureConfig) error {
+	cluster, err := r.getClusterForAzureConfig(ctx, cr)
```

Comment:
Should we handle the case when we don't have a `Cluster` just yet and don't error in that case?